ports-of-call/portable_errors.hpp

```diff
@@ -0,0 +1,198 @@
+  while (true) {
+  }
+#else
+  std::abort();
+#endif // PORTABILITY_STRATEGY_KOKKOS
+}
+} // namespace impl
+
+[[noreturn]] PORTABLE_INLINE_FUNCTION void require(const char *const condition,
+                                                   const char *const message,
+                                                   const char *const filename,
+                                                   int const linenumber) {
+  printf("### ERROR\n  Condition:   %s\n  Message:     %s\n  File:        "
+         "%s\n  Line number: %i\n",
+         condition, message, filename, linenumber);
+  impl::abort(message);
+}
+
+inline void require(const char *const condition, std::string const &message,
+                    const char *const filename, int const linenumber) {
+  require(condition, message.c_str(), filename, linenumber);
+}
+
+inline void require(const char *const condition,
+                    std::stringstream const &message,
+                    const char *const filename, int const linenumber) {
+  require(condition, message.str().c_str(), filename, linenumber);
+}
+
+inline void require_throws(const char *const condition,
+                           const char *const message,
+                           const char *const filename, int const linenumber) {
+  std::stringstream msg;
+  msg << "### ERROR\n  Condition:   " << condition
+      << "\n  Message:     " << message << "\n  File:        " << filename
+      << "\n  Line number: " << linenumber << std::endl;
+  throw std::runtime_error(msg.str().c_str());
+}
+
+inline void require_throws(const char *const condition,
+                           std::string const &message,
+                           const char *const filename, int const linenumber) {
+  require_throws(condition, message.c_str(), filename, linenumber);
+}
+
+inline void require_throws(const char *const condition,
+                           std::stringstream const &message,
+                           const char *const filename, int const linenumber) {
+  require_throws(condition, message.str().c_str(), filename, linenumber);
+}
+
+[[noreturn]] PORTABLE_INLINE_FUNCTION void abort(const char *const message,
+                                                 const char *const filename,
+                                                 int const linenumber) {
+  printf("### ERROR\n  Message:     %s\n  File:        %s\n  Line number: %i\n",
+         message, filename, linenumber);
+  impl::abort(message);
+}
+
+[[noreturn]] inline void abort(std::stringstream const &message,
+                               const char *const filename,
+                               int const linenumber) {
+  abort(message.str().c_str(), filename, linenumber);
+}
+
+[[noreturn]] inline void abort(std::string const &message,
+                               const char *const filename,
+                               int const linenumber) {
+  abort(message.str().c_str(), filename, linenumber);
+}
```

Comment:
Why not template on the type of `msg` rather than overloading on explicit types?